### ISSUE TITLE
Consider to store class in instance, store parent classes in derived class, support run-time method lookup and MRO

### ISSUE THREAD

**Issue by TengMCing** [TengMCing/bandicoot#3]
The current implementation of the class system only supports build-time method lookups, methods are static once the instance is built. 

Later changes to the class by default will not affect the existing instances unless the method body explicitly calls the method of the class. This can be changed by storing the classes in the instance and redirecting all traffic to the class. One issue with this approach is if two instances that share the same class are saved with the command `saveRDS()` and later loaded with the command `readRDS()`, they will point to two different environments with identical content. This means changes to the methods of a class may not be recognized by the instances.

Moreover, modifying the class definition is not recommended by the community in dynamic languages like Python. They call it "monkey patching" given the fact that it may violate the principle of least surprise.

Furthermore, run-time method lookup is more difficult to implement (in the sense of getting it correct) and a little bit slower than build-time class lookups when a method is called. It does have the benefit of saving some memory.

However, storing the classes in instance and storing parent classes in the derived class could be considered, because some methods are extensions to the parent methods. Calling methods using internal information mitigates the risk of calling a non-existing function. On the other hand, staying with the current design forces the user to reproduce the dependencies, which largely avoids the "different environments" problem described above.

In conclusion, the current design of the system is simple enough that the behaviour of the system can be easily predicted. Implementing a more python-like OOP system may lead to difficult and non-trivial bug hunts in edge cases while the benefit to general users provided by the enhanced features may not be obvious. These features will still be considered and will be experimented with in the `experimental` branch.

**Comment by TengMCing** [comment]
MRO has been added to the system (#11). The next step will be to implement dynamic lookup. Before that, a proper design needs to be made for saving and loading objects.

**Comment by TengMCing** [comment]
It is possible that dynamic dispatch will slow down the system a lot considering the overhead. Implementing dynamic dispatch correctly (matching what Python does) is also tricky. We have to manage all the environments properly and not create confusion for general R users.

I think it is not the correct time to fully rewrite the system to implement dynamic dispatch. However, we can consider implementing dynamic dispatch using the current infrastructure (super-like call and dynamic attribute lookup via overloading `$`), but we still need to figure out how to correctly resolve the environment issue.
